### ISSUE TITLE
Restore mock.module using mock.restore not work as expect

### ISSUE THREAD

**Issue by yukikwi** [oven-sh/bun#7823]
### What version of Bun is running?

1.0.20+09d51486e

### What platform is your computer?

Linux 6.5.0-14-generic x86_64 x86_64

### What steps can reproduce the bug?

I create test file with this code inside
```javascript
import axios from "axios"
import { describe, expect, it, mock } from "bun:test"

describe("topic", () => {
  it("test case 1", async () => {
    // before mock
    expect((await axios("https://google.com/", { method: "GET" })).status).toBe(200)

    // mock
    mock.module("axios", () => ({
      default: () => {
        return {
          status: 500
        } 
      }
    }));

    // after mock
    expect((await axios("https://google.com/", { method: "GET" })).status).toBe(500)

    // restore mock & test again
    mock.restore()
    expect((await axios("https://google.com/", { method: "GET" })).status).toBe(200)
  })
})
```

### What is the expected behavior?

I expect this test case must pass

### What do you see instead?

// before mock
expect((await axios("https://google.com/", { method: "GET" })).status).toBe(200) -> this pass

// after mock
expect((await axios("https://google.com/", { method: "GET" })).status).toBe(500) -> this pass

// restore mock & test again
expect((await axios("https://google.com/", { method: "GET" })).status).toBe(200) -> this fail (value is 500)

### Additional information

This one work like what I expected but I think it is not good solution

```javascript
import axios from "axios"
import path from "path"
import { describe, expect, it, mock } from "bun:test"

describe("topic", () => {
  it("test case 1", async () => {
    // before mock
    expect((await axios("https://google.com/", { method: "GET" })).status).toBe(200)

    // mock
    const axiosBak = axios  <-- add this
    mock.module("axios", () => ({
      default: () => {
        return {
          status: 500
        } 
      }
    }));

    // after mock
    expect((await axios("https://google.com/", { method: "GET" })).status).toBe(500)

    // restore mock & test again
    mock.restore()
    mock.module("axios", () => axiosBak) <-- add this
    expect((await axios("https://google.com/", { method: "GET" })).status).toBe(200)
  })
})
```

**Comment by chlorophant** [comment]
Not only is the restore not working, modules that are mocked will conflict with other testfiles that also need to mock the given module. Not sure what the path forward is for this but its a major blocker

**Comment by jpasquers** [comment]
This is the workaround I have been using in the meantime:

```
import { mock } from "bun:test";

export type MockResult = {
    clear: () => void;
};

/**
 *
 * @param modulePath - the path starting from this files' path.
 * @param renderMocks - function to generate mocks (by their named or default exports)
 * @returns an object
 */
export const mockModule = async (
    modulePath,
    renderMocks: () => Record<string, any>,
): Promise<MockResult> => {
    let original = {
        ...(await import(modulePath)),
    };
    let mocks = renderMocks();
    let result = {
        ...original,
        ...mocks,
    };
    mock.module(modulePath, () => result);
    return {
        clear: () => {
            mock.module(modulePath, () => original);
        },
    };
};
```

Then for each test suite declare

```
    let mocks: MockResult[] = [];
    afterEach(() => {
        mocks.forEach((mockResult) => mockResult.clear());
        mocks = [];
    });
```

Then finally for example in the test

```
        mocks.push(
            await mockModule("./utils/aws", () => ({
                ensureAwsSsoSession: jest.fn(() => Promise.resolve()),
            })),
        );
```

I can't stress enough that this is a hacky unstable solution that I am only leveraging because our unit tests would be dead in the water otherwise. Hopefully this helps others until a permanent solution comes through 👍

**Comment by Jarred-Sumner** [comment]
As a temporary workaround, you can try `jest.restoreAllMocks`

```
import {jest} from 'bun:test';

jest.restoreAllMocks();
```

**Comment by jpasquers** [comment]
@Jarred-Sumner hmm unfortunately that didn't resolve it for me, or at the very least the way I used it wasn't working. I leveraged it as an `afterEach`, i.e.

```
afterEach(() => jest.restoreAllMocks());
```

And using the standard `mock.module` (example usage):

```
        mock.module("./utils/aws", () => ({
            ensureAwsSsoSession: jest.fn(() => Promise.resolve()),
        }));
```

**Comment by septemptus** [comment]
> As a temporary workaround, you can try `jest.restoreAllMocks`
> 
> ```
> import {jest} from 'bun:test';
> 
> jest.restoreAllMocks();
> ```

This does work for me, but the import line is required -- if the `jest` global variable is used then it doesn't work

**Comment by cuquo** [comment]
I'm facing the same issue with mock.module, I also tried mocking the module with `require.resolve('path')` in order to see if that would create another mock for the same module but with no luck.

Ideally the mock.module should only be valid inside that test, other tests could mock the same module with other values like jest. Global mocks should be added in the preload file IMHO.

**Comment by RichAyotte** [comment]
> As a temporary workaround, you can try `jest.restoreAllMocks`
> 
> ```
> import {jest} from 'bun:test';
> 
> jest.restoreAllMocks();
> ```

This doesn't work. Any other workarounds other than instantiating a new bun process for every test file?

**Comment by MonsterDeveloper** [comment]
Since #10210 is closed now, we really need the fix for `mock.restore` or `jest.restoreAllMocks`… Right now it’s really hard to use bun’s test runner on more or less big project with module mocking without spawning separate process for every test file.

**Comment by RobinTail** [comment]
I think `mock.module()` should return something, having `.mockRestore()` method.

**Comment by dmvvilela** [comment]
Any updates on this? When i mock a common module it just breaks everything on other tests.

**Comment by ziroock** [comment]
I am facing the same issue! Tests break once I mock a function A in a unit test and then use the same function A in an integration test. I have separate files for the unit and integration tests. Mocking a module mocks globally, not on describe() level. Are there any updates?

**Comment by ziroock** [comment]
I am facing the same issue! Tests break once I mock a function A in a unit test and then use the same function A in an integration test. I have separate files for the unit and integration tests. Mocking a module mocks globally, not on describe() level. I used mock.restore() and jest.restoreAllMocks(), but still nothing. Are there any updates?

**Comment by aryzing** [comment]
Seems the conversation has shifted a bit towards cross-test module mocking, here's a related issue on that: https://github.com/oven-sh/bun/issues/12823

**Comment by jstlaurent** [comment]
I ended up using @jpasquers' solution, from this earlier [comment](https://github.com/oven-sh/bun/issues/7823#issuecomment-1908445634). It's unfortunate that it doesn't work out of the box, but the benefits I get from testing outweigh the additional complexity.

I tweaked the solution to use a class, so it's all encapsulated in one place. It looks like this:
```typescript
/**
 * Due to an issue with Bun (https://github.com/oven-sh/bun/issues/7823), we need to manually restore mocked modules
 * after we're done. We do this by setting the mocked value to the original module.
 *
 * When setting up a test that will mock a module, the block should add this:
 * const moduleMocker = new ModuleMocker()
 *
 * afterEach(() => {
 *   moduleMocker.clear()
 * })
 *
 * When a test mocks a module, it should do it this way:
 *
 * await moduleMocker.mock('@/services/token.ts', () => ({
 *   getBucketToken: mock(() => {
 *     throw new Error('Unexpected error')
 *   })
 * }))
 *
 */
export class ModuleMocker {
  private mocks: MockResult[] = []

  async mock(modulePath: string, renderMocks: () => Record<string, any>) {
    let original = {
      ...(await import(modulePath))
    }
    let mocks = renderMocks()
    let result = {
      ...original,
      ...mocks
    }
    mock.module(modulePath, () => result)

    this.mocks.push({
      clear: () => {
        mock.module(modulePath, () => original)
      }
    })
  }

  clear() {
    this.mocks.forEach(mockResult => mockResult.clear())
    this.mocks = []
  }
}
```

**Comment by RobinTail** [comment]
The solution would be making `mock.restore()` work in `bun`.

**Comment by MonsterDeveloper** [comment]
@jstlaurent the main issue with this approach is that you’re still importing the original module. In many cases mocking is used for the whole module in order to not import it itself or its dependencies.

So right now it’s not possible to restore the module mock in Bun without importing the original module. 

Since Bun seems to spawn processes quite fast, I’ve come to love this approach: 

```sh
find . -name \"*.test.ts\" | xargs -I {} sh -c 'bun test {} || exit 255'
```

It runs each test file in its own process sequentially and therefore doesn’t come across issues related to module mocking not restoring. It also stops the execution if one of the test files failed.

It’s also possible to increase test run speed by running `xargs` processes in parallel, yet I found that even sequential runs finish all tests in around 15 seconds (~6k lines of tests).

**Comment by juriadams** [comment]
Unfortunately, none of the workarounds mentioned above resolved this blocker for me. 

While the implementation is pending, perhaps the documentation for [**Module mocks with `mock.module`**](https://bun.sh/docs/test/mocks#module-mocks-with-mock-module) could be updated to include a warning banner indicating that restoring module mocks is currently not possible.

**Comment by redbmk** [comment]
It's really weird, for me `mock.restore()` seems to be working fine locally, but in CI mocked modules between different test files will conflict with each other. I feel like there must be some environment variable causing different behavior but I'm not able to track it down yet.

I'm using 1.1.39 in both environments

**Comment by notsoluckycharm** [comment]
I agree with the conversation, but I wanted to share my way of doing this which doesn't require you calling the .clear method. 

```typescript
import { mock } from "bun:test";

/**
 *
 * @param modulePath - the path starting from this files' path.
 * @param renderMocks - function to generate mocks (by their named or default exports)
 * @returns an object
 */
export const mockModule = async (
  modulePath: string,
  renderMocks: () => Record<string, any>,
) => {
  let original = {
    ...(await import(modulePath)),
  };
  let mocks = renderMocks();
  let result = {
    ...original,
    ...mocks,
  };
  mock.module(modulePath, () => result);
  return {
    [Symbol.dispose]: () => {
      mock.module(modulePath, () => original);
    },
  };
};
```

This takes advantage of the Typescript `using` keyword - pretty cool! 

```typescript
    using parseAndRollMock = await mockModule('../commands/utils/parseAndRoll', () => ({
      parseAndRoll: jest.fn().mockReturnValue({ total: 15, details: [15] }),
    }));
```

No need to call clear or use before/after hooks for this. As soon as it leaves scope, it will call the reset methods defined in Symbol.dispose - credit due to @jpasquers since I borrowed it from there and just expanded on it.

**Comment by platzhersh** [comment]
@notsoluckycharm that's a pretty neat fix for now 🙌  How would you make assertions like `.toHaveBeenCalledWith` with this?

**Comment by Chr1d5ter** [comment]
> > As a temporary workaround, you can try `jest.restoreAllMocks`
> > ```
> > import {jest} from 'bun:test';
> > 
> > jest.restoreAllMocks();
> > ```
> 
> This doesn't work. Any other workarounds other than instantiating a new bun process for every test file?

The only quick ( and very very dirty ) way I got around this was to...

`
import * as WeightModule from "./modules/weights";

describe("Weight Module", () => {
  const originalModule = JSON.parse(JSON.stringify(WeightModule));

  beforeAll(() => {
    mock.module("./modules/weights", () => ({
      weightData: weightTestData,
    }));
  });

  afterAll(() => {
    mock.restore();
    mock.module("./modules/weights", () => originalModule);
  });

{Tests go here...}
});
`
Hope this helps

**Comment by tpetry** [comment]
For some reason the workarounds shared here didn't work for me. In some cases the a mock was still not reset. After a lot of trying I gave up.

I am now using the following command to run every test in a new bun process. So I've implemented test isolation that way:

```sh
find . -name "*.test.ts" -exec sh -c 'bun test "$1" || kill "$PPID"' sh {} \;
```

**Comment by Bernhard-Kodio** [comment]
I just ran into this issue. I though I was going crazy. My tests are breaking when I try to run them all in one go, but they pass with flying colors when I run each test file individually.

I was hoping to use module mocking to greatly reduce the amount of boilerplate in a project, so this is really hurting the usability of bun.

I've tried the workarounds mentioned in this thread but none of them work. The only solution is to use a script that scans for test files and then runs bun test on each file individually.

It's also quite frightening that a critical bug like this has apparently been left unfixed for several years.

**Comment by jboyesreach** [comment]
FYI, the workaround is working for me, where you:
1.  import the thing to be mocked
2. take a backup copy of it
3. mock.module(....
4. in afterAll, do mock.module(... again to restore the backup.

but I would LOVE to be able to rip that ugly crap out of all my tests.

**Comment by MonsterDeveloper** [comment]
Happy to see more and more people bringing this to attention.

While I love Bun's new cool features and native replacements for databases, utilities, etc., it feels like this issue, which is far more crucial, is either not on the team's radar or is being deliberately ignored in favor of shipping "new cool stuff" instead of addressing "boring bug fixes."

I really hope it's the former, so if anyone has contact with the core team developers, we'd really appreciate it if they could bring this to their attention.

---

Personally, I'm using `find` to run each test file separately. This fixes the bug but is a lot slower than Vite and even Jest. Also, it's not really an option to run these tests in a Node environment, since we're using Bun as a runtime too.

The workaround of creating a mocked module registry in JS is a good idea, but it's not an option either, as it still imports the original module—which is exactly what we're trying to avoid.

**Comment by jboyesreach** [comment]
There IS an open PR: https://github.com/oven-sh/bun/pull/18171
But it's very quiet over there. Nothing's happened since March (2025).
Bun seems to get releases almost every week, but this issue never gets any love.

**Comment by Bernhard-Kodio** [comment]
> FYI, the workaround is working for me, where you:
> 
>     1. import the thing to be mocked
> 
>     2. take a backup copy of it
> 
>     3. mock.module(....
> 
>     4. in afterAll, do mock.module(... again to restore the backup.
> 
> 
> but I would LOVE to be able to rip that ugly crap out of all my tests.

Importing a module can have side effects, so you need to be a bit careful with a workaround like that.

**Comment by pirog** [comment]
Here is another low effort way to do this while the underlying isolation/restore stuff is being worked on.

```js
import { describe, test, expect } from 'bun:test';

// https://github.com/oven-sh/bun/issues/7823
import functionPreviouslyMocked from 'functionPreviouslyMocked.js?AVOID_PREVIOUS_MOCK';

describe('functionPreviouslyMocked', () => {
  test('should do something', () => {
    expect(functionPreviouslyMocked()).toBe('doing something);
  });
});
```

adding `?SOMETHINGELSE` to the import string will load a fresh copy of the module and avoid the mocked version that is cached.

**Comment by RageLtd** [comment]
In case anyone was still struggling with this, I created a [package of tools](https://github.com/RageLtd/bun-test-utils) to help with the mock isolation issues. Use it, or don't 🤷  Just figured I'd share since it was bothering me to have to avoid `bun:test` - `vitest` just feels so slow now...

**Comment by gel1123** [comment]
> Not only is the restore not working, modules that are mocked will conflict with other testfiles that also need to mock the given module. Not sure what the path forward is for this but its a major blocker

Could it be that this issue doesn't occur on macOS?  
I remember that module mocks were properly reset between test files on macOS.  
Unfortunately, I can't verify it myself right now due to some circumstances,  
but has anyone experienced this issue *on macOS* as well?

**Comment by dimfeld** [comment]
So long as you aren't worried about side effects from importing modules, I've had good success with the "ModuleMocker" class described in https://github.com/oven-sh/bun/issues/7823#issuecomment-2328937607. 

I tweaked my copy it a bit so that if you mock a module twice without calling `clear`, it still correctly restores the original, but otherwise it has completely solved the problem for me while we wait for Bun to fix the problem properly.

**Comment by gentlementlegen** [comment]
According to the docs:

**mock.clearAllMocks**
> Reset all mock function state (calls, results, etc.) without restoring their original implementation.


https://bun.com/docs/test/mocks#clear-all-mocks-with-mock-clearallmocks

**mock.restore**
> Instead of manually restoring each mock individually with mockFn.mockRestore(), restore all mocks with one command by calling mock.restore(). Doing so does not reset the value of modules overridden with mock.module().

https://bun.com/docs/test/mocks#restore-all-function-mocks-with-mock-restore

After bumping my head on the table for a while, I figured out that I could use `spyOn` instead to mock! So basically, the changes are:

```typescript
mock.module("./my-module", () => true);
```

would become

```typescript
spyOn(await import("./my-module")).mockReturnValue(true);
```

And I call `mock.restore` and `mock.ClearAllMocks` within the `beforeEach` of the tests. So far, it seems that this technique makes the mock not persist within different tests.

**Comment by RageLtd** [comment]
I was so hopeful for 1.2.20, but no dice 😢

**Comment by npearson72** [comment]
Guess this is the price of living on the edge 😆 

Just chiming in as yet another hopeful dude waiting on this getting resolved. I'm hooked on that bun speed man.

**Comment by jboyesreach** [comment]
> > > Not only is the restore not working, modules that are mocked will conflict with other testfiles that also need to mock the given module. Not sure what the path forward is for this but its a major blocker
> > 
> > 
> > Could it be that this issue doesn't occur on macOS? I remember that module mocks were properly reset between test files on macOS. Unfortunately, I can't verify it myself right now due to some circumstances, but has anyone experienced this issue _on macOS_ as well?
> 
> I do have a similar behaviour where the tests passes fine on multiple macOS machines but not in CI with the same repo setup

I've seen it happen where, because the test suites are run in alphabetical order, and different OS's handle that differently, on one OS maybe the suite that mocks the module runs AFTER the one that needs the real module so everything is fine, but on other OS's they run in a different order and badness happens.

This new PR https://github.com/oven-sh/bun/pull/21826 is rather exciting but I don't want to get my hopes up after the previous one was rejected.

**Comment by ialexryan** [comment]
Currently laying the groundwork to convince my team to adopt bun, but I ran into this on day 1 and it's a hard stop. I can't go to them and ask them to adopt a tool where test execution is nondeterministic, and outcomes depend on which files are tested in which order. @Jarred-Sumner is this on the roadmap to be fixed?

**Comment by jboyesreach00** [comment]
Any progress or update on this? This is a blocker for migrating existing projects to Bun.

**Comment by giavinh79** [comment]
> I found this workaround works well (in v1.3.5):
> 
> _module.ts_
> 
> export function fn() {
>   return true
> }
> 
> export default fn
> _module.test.ts_
> 
> import { describe, expect, it, mock } from 'bun:test'
> // Use `import * as ...` to handle the different export types (named, default, or others) of modules
> import * as alias from './module'
> 
> // Make a copy before you mock it.
> // If you want a deep copy, just use deep clone instead of object spread.
> const originalModule = { ...alias }
> 
> describe('test', () => {
>   it('should mock and re-mock correctly', () => {
>     expect(alias.fn()).toEqual(true)
> 
>     // Mock
>     mock.module('./module', () => ({
>       ...alias,
>       fn: () => false,
>     }))
>     expect(alias.fn()).toEqual(false)
> 
>     // Re-mock, use the copy too
>     mock.module('./module', () => ({ ...originalModule }))
>     expect(alias.fn()).toEqual(true)
>   })
> })
> This also works on external modules!

Would this be fragile under parallel test execution & race conditions? With more complex examples (i.e. more work between the mock & re-mock), couldn't other tests run during that window and thus be affected by that global mock?

**Comment by jboyesreach** [comment]
I wonder is this not being prioritised because "serious devs use dependency injection, instead of dependency mocking" ???

**Comment by jff1625** [comment]
> Partially addressed by [#29354](https://github.com/oven-sh/bun/pull/29354) — `bun test --isolate` (experimental) gives each test file a fresh `JSGlobalObject`, so `mock.module` registrations from one file don't carry into the next. That removes the need for cross-file restore.
> 
> `mock.restore()` semantics within a single file are unchanged; this issue stays open for that.

it's a LOT slower tho.

```
bun test
Ran 262 tests across 26 files. [5.00s]

bun test --isolate
Ran 262 tests across 26 files. [89.31s]

bun test --isolate --parallel
Ran 262 tests across 26 files. [38.63s]
```
